### PR TITLE
add deeper query options and check_connectivity function

### PR DESCRIPTION
**Pull request by daikman**
I've added the ability to filter rows and columns using the 'q' and 'fields' parameters of the datastore_search endpoint, respectively.

I also added a function that checks the connection to CKAN in order to give better error messages from `get_resource` and `get_dataset`. And for `get_resource` I added some extra error handling.

Made small changes to documentation of functions too.

### CODE REVIEW

**Comment by csillasch** [comment]
I don`t think we should set the query limit to 100 if there is a filtering query with a larger response. 
As a user I would find it tedious to get a warning message instructing me to repeat the query specifying the number of rows, rather than getting that response in the first place - less clunky and frustrating from user perspective.
Did you have specific reasoning to include the limit?

**Comment by csillasch** [comment]
Also looked at the failed checks - think these will be resolved by updating the tests to expect the new error messages, but if not may need a bit more attention

**Comment by daikman** [comment]
> I don`t think we should set the query limit to 100 if there is a filtering query with a larger response. As a user I would find it tedious to get a warning message instructing me to repeat the query specifying the number of rows, rather than getting that response in the first place - less clunky and frustrating from user perspective. Did you have specific reasoning to include the limit?

If there is no limit set in a datastore_search query, then it will return 100 results by default. I thought it would be good to make that explicit. But you are right that it would be annoying for users! 

Instead, I've changed it so that if they don't set a limit themselves, it becomes 99999. If a query matches more rows than that, the warning tells them to download the whole resource and apply filters afterwards. If their query matches 99999 rows or less, then the user won't notice that this limit has been set.

> Also looked at the failed checks - think these will be resolved by updating the tests to expect the new error messages, but if not may need a bit more attention

This is a good point about tests! Even if this isn't the reason for failed checks I should update them anyway.

**Comment by csillasch** [comment]
> If there is no limit set in a datastore_search query, then it will return 100 results by default. I thought it would be good to make that explicit. But you are right that it would be annoying for users!
> 
> Instead, I've changed it so that if they don't set a limit themselves, it becomes 99999. If a query matches more rows than that, the warning tells them to download the whole resource and apply filters afterwards. If their query matches 99999 rows or less, then the user won't notice that this limit has been set.
> 

Good point - that sounds like a sensible solution. I think previously we used a single-row call first to determine the exact row number to return.

**Review by Moohan — COMMENTED** [review]
Some great additions! The filtering seems like a nice interface, well done with that 👍

Most of my comments are around trying to limit the number of API calls. I think you're right that the error messaging needs to be better to explain how/why the query failed but I think this can be done without adding lots of extra calls.

Other comments are similar to Csilla i.e. what to do when a user doesn't specify a limit on row number. I would always lean to returning the full extract in this case, as a user I think I'd expect that, this does mean extra work (and probably calls) are required in the background.

I think this would combine very nicely with your batch download PR #6, for example, if someone specifies filtering on a very large extract, instead of just returning the full data unfiltered we could fall back to the batch function to eventually return the full filtered extract.

My other suggestion would be to have a look at `rlang` for the errors and warnings: [tidyverse style guide - error messages](https://style.tidyverse.org/error-messages.html). They allow automatically formatting errors really nicely, especially with the latest [1.0.0 release](https://www.tidyverse.org/blog/2021/12/rlang-1-0-0-errors/)

**Comment by daikman** [comment]
I've made three major changes to this pull request:

Refactoring the way the package makes API calls in order to better handle errors and reduce unnecessary API calls. 
Using {cli} to throw and format error messages.
I also separated all functions into their own files and added tests for each.

Forgot to add any of your suggested changes @Moohan before I made these changes! But thanks for your comments they were very helpful

File: R/phs_GET.R [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+  response <- httr::GET(
+    url,
+    user_agent = httr::user_agent(
+      "https://github.com/Public-Health-Scotland/phsmethods"
+    )
+  )
+
+  # Check for response from server
+  is_response <- httr:::is.response(response)
+  if (!is_response)
```
Comment (Moohan) [inline R/phs_GET.R:23]: It's generally bad to use unexported functions `:::` if you check the httr source for this function it's actually super simple anyway: [https://github.com/r-lib/httr/blob/main/R/response.r](https://github.com/r-lib/httr/blob/main/R/response.r)

```suggestion
  if (!inherits(response, "response"))
```

File: R/get_resource.R [hunk 2]
```diff
@@ -13,108 +15,70 @@
...
+      the whole resource and apply filters/selections locally."
+    ))
+
+  # if more rows were requested than received
+  # let the user know
+  if (!is.null(rows) && query$limit > total_rows)
+    cli::cli_warn(c(
+      "You set {.var rows} to {query$limit} but
+      only {total_rows} rows matched your query."
+    ))
```
Comment (Moohan) [inline R/get_resource.R:73]: Does this need to be a warning? `cli_alert_warning()` is nicely formatted like a warning but is actually just a message.
```suggestion
    cli::cli_alert_warning(c(
      "You set {.var rows} to {query$limit} but
      only {total_rows} rows matched your query."
    ))
```

File: R/parse_row_filters.R [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+    cli::cli_abort(c(
+      "Invalid input for {.var row_filters}",
+      i = "{names(row_filters)[which(too_many)]} in {.var row_filters} has too many values. ",
+      x = "The {.var row_filters} list must only contain vectors of length 1."
+    ))
+
+  # check if any items in the list/vector have the same name
+  unique_names <- length(
+    unique(names(row_filters))) == length(names(row_filters)
+  )
```
Comment (Moohan) [inline R/parse_row_filters.R:24]:
```suggestion
  unique_names <- length(unique(names(row_filters))) == length(names(row_filters))
```
Spread out makes this hard to interpret. I initially thought the brackets were wrong.

**Review by Moohan — COMMENTED** [review]
I don't have time to do a proper review at the moment so don't wait for me! I did just build it and had a little play and it seems to work well.

There were two build notes, one on the usage of `:::` and the other on 'non-standard' files, those files should either be added to the `.Rbuildignore` or just removed (do we need the travis.yml?)

Other comments would be to add a news / readme update and increment the version before merging all of this 👍

**Comment by csillasch** [comment]
Would it be helpful to have a message also when dump is being used for row>99999 requests but there is no query/filtering? Just to avoid any confusion when the returned number of rows doesn`t match the rows requested.

File: R/parse_row_filters.R (hunk 3 above)
Comment (daikman) [reply R/parse_row_filters.R:24]: You're right! I've decided to split it up into separate lines to make it extra clear:

```
# check if any items in the list/vector have the same name
# find number of unique names
n_u_row_filters <- length(unique(names(row_filters)))
# find total number of names
n_row_filters <- length(names(row_filters))
# if same, all names are unique
unique_names <- n_u_row_filters == n_row_filters
```

**Pull request merged**